beetmoverscript/src/beetmoverscript/script.py

```diff
@@ -285,6 +290,206 @@ async def upload_data(context):
+            if map_["taskId"] != taskId:
+                continue
+
+            for artifact in artifacts:
+                # the artifact path with the scriptworker-specific leading
+                # directories removed. ie: in the same form it is in the task
+                # it was pulled from
+                leading_dir = os.path.join(work_dir, "cot", taskId) + "/"
+                if not artifact.startswith(leading_dir):
+                    raise ScriptWorkerTaskException("cannot determine relative artifact path")
+
+                relative_artifact = artifact[len(leading_dir):]
+
+                destinations = []
+                # We need to look at non-'*' paths separate from '*' paths.
+
+                for input_path, output in other_paths.items():
+                    # Skip any input paths that don't match the artifact name.
+                    if not fnmatch.fnmatch(artifact, input_path):
+                        continue
+
+                    # If there's already destinations, we've already seen this artifact,
+                    # and we have a clash.
+                    if destinations:
+                        errors.append(f"'{relative_artifact}' matched multiple concrete paths")
+                    else:
+                        destinations.extend(output["destinations"])
+
+                # If we didn't find any destinations for the artifact above, and
+                # there are full glob destinations, it should go there.
+                if full_glob_destinations and not destinations:
+                    destinations.extend(full_glob_destinations)
+
+                # If there are destinations, create a fully concrete entry,
+                # mapping the full path to the file on disk to destinations
+                # it belongs, with any requested prefixes removed.
+                if destinations:
+                    if strip_prefixes:
+                        dest_artifact = remove_prefixes(relative_artifact, strip_prefixes)
+                    else:
+                        dest_artifact = relative_artifact
+
+                    concretePaths[artifact] = {"destinations": []}
+                    for d in destinations:
+                        if d.endswith("/"):
+                            d = os.path.join(d, dest_artifact)
+                        concretePaths[artifact]["destinations"].append(d)
+
+        concreteArtifactMap.append(
+            {
+                "paths": concretePaths,
+                "taskId": map_["taskId"],
+            }
+        )
+
+    log.info("Found concrete artifact map:")
+    log.info(json.dumps(concreteArtifactMap, indent=2))
+
+    ensure_no_overwrites_in_artifact_map(concreteArtifactMap)
+
+    if errors:
+        raise ScriptWorkerTaskException(*errors)
+
+    return concreteArtifactMap
+
+
+def remove_prefixes(string, prefixes):
+    for p in prefixes:
+        if string.startswith(p):
+            string = string[len(p):]
```

Comment:
nit: could return here unless there's a case where we want to strip multiple prefixes from the same path?